File: DP/300_Longest_Increasing_Subsequence/java/main.java
class Solution {
  private int[] memo;
  public int lengthOfLIS(int[] nums) {
      if (nums.length == 1) {
          return 1;
      }
      
      memo = new int[nums.length];
      
      int max = Integer.MIN_VALUE;
      for(int i=nums.length-1; i>=0; i--) {
          int curMax = findMaxSubsequence(i, nums);
          if (max < curMax) {
              max = curMax;
          }
      }
      
      return max;
  }
  
  private int findMaxSubsequence(int index, int[] nums) {
      // 若是最尾巴的那隻 長度只有1
      if (index == nums.length-1) {
          this.memo[index] = 1;
      }
      
      if (this.memo[index] == 0) {
          // 預設先給1
          this.memo[index] = 1;
          // 計算目前index的最大subsequece有多少
          for (int i=index+1; i<nums.length; i++) {
              if (nums[index] < nums[i]) {
                  // 找到最近已經有算出他的max sub的lentgh
                  
                  // 且若在後面遇到比自己大的，檢查目前該值是否比之前找到的還大
                  if (this.memo[index] <= this.memo[i]) {
                     this.memo[index] = this.memo[i] + 1;
                  }
              }
          }            
      } 
      
      return this.memo[index];
  }
}

File: DP/300_Longest_Increasing_Subsequence/golang/main.go
func lengthOfLIS(nums []int) int {
	if len(nums) == 1 {
		return 1
	}

	memo := make([]int, 2500)

	max := -1000000 //   -10^4 < nums[i] < 10^4

	for i := len(nums) - 1; i >= 0; i-- {
		curMax := findMaxSubsequence(i, nums, memo)
		if max < curMax {
			max = curMax
		}
	}

	return max
}

func findMaxSubsequence(index int, nums []int, memo []int) int {
	// 若是nums最末端，長度只有1
	if index == len(nums)-1 {
		memo[index] = 1
	}

	// 如果當前的index沒有計算longest increasing subsequence
	if memo[index] == 0 {
		// 設置初始值為1 (自己的長度)
		memo[index] = 1

		for i := index + 1; i < len(nums); i++ {
			// 找到比目前index的值，更大的對象，才能夠納入自己的subsequence
			if nums[index] < nums[i] {
				// 若找到比自己更大的longest increasing subsequence，進行更新
				if memo[index] <= memo[i] {
					memo[index] = memo[i] + 1
				}
			}
		}
	}
	return memo[index]
}